packages/vx-demo/src/sandboxes/vx-tooltip/Example.tsx

```diff
@@ -0,0 +1,244 @@
+import React, { useState, useCallback } from 'react';
+import useMeasure from 'react-use-measure';
+import { Tooltip, TooltipWithBounds, useTooltip, defaultStyles, Portal } from '@vx/tooltip/src';
+
+export type TooltipProps = {
+  width: number;
+  height: number;
+  showControls?: boolean;
+};
+
+interface TooltipData {
+  text: string;
+  containerX: number;
+  containerY: number;
+  pageX: number;
+  pageY: number;
+}
+
+const positionIndicatorSize = 8;
+
+const tooltipStyles = {
+  ...defaultStyles,
+  backgroundColor: 'rgba(53,71,125,0.8)',
+  color: 'white',
+  width: 152,
+  height: 72,
+  padding: 12,
+};
+
+export default function Example({ width, height, showControls = true }: TooltipProps) {
+  // bounds of the container are needed to convert page coordinates to container coordinates
+  const [ref, ownBounds] = useMeasure({ scroll: true });
```

Comment:
Do end users have to import `useMeasure` themselves and copy all these boilerplate code in order to use the portal mode?

Any plan to integrate `renderInPortal` as an option for `useTooltip`?

Reply:
currently, yes, users must use `useMeasure`, but if you have thoughts on how to abstract this I haven't given it a lot of thought. a couple points:

- `vx` is low-level and the components are meant to be very generic. So I think some work on the user's end to use this is okay as we are _just_ adding the functionality, but hopefully we can make it a little easier.
  - also noting that I am adding this as part of the `XYChart` package, which will abstract everything for the developer

- It's a little tricky to abstract all of the coordinate logic because there are different use cases. For example, _this_ usage results in the `Tootlip` exactly following the mouse cursor which uses `pageX/Y`. But if you want the `Tootlip` to snap to data points you can't use `pageX`, you would have to
1) find the closest `datum` 
2) scale the `x` and `y` of the `datum`, which gives some `px` values in `svg` coordinates 
3) convert the `svg` coordinates  to `page` coordinates for use in the tooltip

- `useTooltip` is very generic and only holds `Tooltip` state; it doesn't do any rendering so it can't handle the `Portal` wrapping on its own. `Tooltip` / `TooltipWithBounds` might be a better candidate as it actually does rendering (see next point)

- perhaps we could create something like `const [coordinateRef, Tooltip] = useTooltipInPortal()` where you could pass `ref={coordinateRef}` to whatever container that will match the coordinates you pass to `<Tooltip left={..} top={..}>{content}</Tooltip>` (this would support following the cursor position, or adding extra logic like snapping to data points). `Tooltip` could then handle conversion of those coordinates to page coordinates using the `bounds` of `coordinateRef` (using `useMeasure` behind the scenes)
 




However, I am adding this for use in I am adding thi

Reply:
Thanks for the additional context.

I don't have a clear idea who are the intended users for the different levels of abstractions, but it simply feels weird to me that we are adding a `Portal` component to `vx/tooltip` but still left a good amount of heavy lifting related to portal outside of the package itself.

`coordinateRef` sounds cool. Haven't dug into it too much, but I'm wondering whether it's possible to do everything in just one hook. Something like this:

```tsx
function useTooltip({
  hideDelay = 300,
  xOffset = 0,
  yOffset = 0,
  portalContainer = document.body,
  getCoordinate,
}) {
  const [measureRef, bounds, forceRefresh] = useMeasure({ scroll: true });
  const tooltipElement = useRef(null);
  const tooltipTimeout = useRef(null);

  const Tooltip = {
    showTooltip(datum, e) {
      const elem = tooltipElement.current;
      if (!elem) return;
      const pos = getCoordinate(datum, e);
      elem.style.left = `${bounds.x + xOffset + pos.x}px`;
      elem.style.right = `${bounds.y + yOffset + pos.y}px`;
    }
  }

  useLayoutEffect(() => {
    const elem = tooltipElement.current;
    if (!portalContainer || !elem) {
        return null;
    }
    const timeout = tooltipTimeout.current;
    portalContainer.appendChild(elem);
    return () => {
      if (timeout) clearTimeout(timeout);
      tooltipTimeout.current = setTimeout(() => {
        if (portalContainer.contains(elem)) {
          portalContainer.removeChild(elem);
        }
      }, hideDelay);
    }
  }, [portalContainer, hideDelay])

  return [measureRef, Tooltip];
}


const [coordinateRef, { showTooltip, hideToolTip }] = useTooltip({
  hideDelay: 300,
  xOffset: 10,
  yOffset: 20,
  getCoordinate: (datum, e) => {
    return {
      x: xScale(datum.x),
      y: yScale(datum.y)
    }
  }
});

<svg ref={coordinateRef}>
  {bars.map(bar => (
    <rect
      onMouseLeave={hideTooltip}
      onMouseMove={showTooltip.bind(null, bar)}
    />
 ))}
</svg>;

```

Reply:
Hmm, I think _not_ having the dev handle rendering of the `Tooltip` at all and having `showTooltip` do this feels a little _too_ magical ,and doesn't fit the declarative nature of `react` as nicely. Also not sure if specifying `portalContainer` will work with SSR since `document` doesn't exist 🤔 

I also don't think we should couple the current `useTooltip` hook (which manages tooltip **_state_**, not rendering) with the `Portal` **_rendering_** logic (which shouldn't care about state) because a consumer might want to use the two separately (e.g., in XYChart I want to get `tooltipData` from `Context`, not the hook).

Working on a separate PR to add the hook I outlined above 🍾